File: solutions/python/perfect-numbers/perfect_numbers.py
from math import sqrt


def classify(number: int) -> str:
    """
    Classifies a given positive integer as "perfect", "deficient", or "abundant"
    based on the sum of its proper divisors excluding the number itself. A number
    is classified as:

    - Perfect: if the sum of its divisors equals the number.
    - Deficient: if the sum of its divisors is less than the number.
    - Abundant: if the sum of its divisors is greater than the number.

    :param number: The positive integer to be classified.
    :type number: int
    :return: A string indicating the classification of the number ("perfect",
        "deficient", or "abundant").
    :rtype: str
    :raises ValueError: If the input number is not a positive integer.
    """
    if number <= 0:
        raise ValueError("Classification is only possible for positive integers.")

    aliquot_sum = (number > 1) + sum(
        sum({i, number // i}) for i in range(2, int(sqrt(number)) + 1) if not number % i
    )
    return "perfect" if aliquot_sum == number else "deficient" if aliquot_sum < number else "abundant"

File: solutions/scala/perfect-numbers/src/main/scala/PerfectNumbers.scala
enum NumberType:
  case Perfect, Abundant, Deficient

object PerfectNumbers:
  def classify(number: Int): Either[String, NumberType] =
    if number <= 0 then
      Left("Classification is only possible for natural numbers.")
    else
      val aliquotSum = sumFactors(number)
      number match
        case 1 => Right(NumberType.Deficient)
        case x if aliquotSum < x => Right(NumberType.Deficient)
        case x if x == aliquotSum => Right(NumberType.Perfect)
        case _ => Right(NumberType.Abundant)

  private def sumFactors(number: Int): Int =
    (2 to Math.sqrt(number).toInt)
      .filter(number % _ == 0)
      .foldLeft(1) { (sum, factor) =>
        val complementaryFactor = number / factor
        if factor != complementaryFactor then
          sum + factor + complementaryFactor
        else
          sum + factor
      }